grpc.go

```diff
@@ -56,40 +58,85 @@ type onlineResource struct {
 
 var emptyResp = &gpb.Empty{}
 
-func onlineCPUMem() error {
-	resourceList := []onlineResource{
-		{
-			sysfsOnlinePath: sysfsCPUOnlinePath,
-			regexpPattern:   cpuRegexpPattern,
-		},
-		{
-			sysfsOnlinePath: sysfsMemOnlinePath,
-			regexpPattern:   memRegexpPattern,
-		},
+var onlineCPUMemMux sync.Mutex
+
+const onlineCPUMemWaitTime = 500 * time.Millisecond
+
+const onlineCPUMaxTries = 5
+
+func onlineResources(resource onlineResource) (uint32, error) {
+	files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	if err != nil {
+		return 0, err
 	}
 
-	for _, resource := range resourceList {
-		files, err := ioutil.ReadDir(resource.sysfsOnlinePath)
+	var count uint32
+	for _, file := range files {
+		matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
 		if err != nil {
-			return err
+			return count, err
 		}
 
-		for _, file := range files {
-			matched, err := regexp.MatchString(resource.regexpPattern, file.Name())
-			if err != nil {
-				return err
-			}
+		if !matched {
+			continue
+		}
 
-			if !matched {
-				continue
-			}
+		onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		status, err := ioutil.ReadFile(onlinePath)
+		if err != nil {
+			// CPU was cold plugged
+			continue
+		}
 
-			onlinePath := filepath.Join(resource.sysfsOnlinePath, file.Name(), "online")
+		if strings.Trim(string(status), "\n\t ") == "0" {
 			ioutil.WriteFile(onlinePath, []byte("1"), 0600)
+			count++
 		}
 	}
 
-	return nil
+	return count, nil
+}
+
+func forceOnlineResources(resource onlineResource, expectedResources, maxTries uint32) error {
+	var count uint32
+	for i := uint32(0); i < maxTries; i++ {
+		r, err := onlineResources(resource)
+		if err != nil {
+			return err
+		}
+		count += r
+		if count == expectedResources {
+			return nil
+		}
+		time.Sleep(onlineCPUMemWaitTime)
+	}
+
+	return fmt.Errorf("only %d of %d were connected", count, expectedResources)
+}
+
+func onlineCPUMem(req *pb.OnlineCPUMemRequest) {
+	onlineCPUMemMux.Lock()
+	defer onlineCPUMemMux.Unlock()
+
+	cpuResource := onlineResource{
+		sysfsOnlinePath: sysfsCPUOnlinePath,
+		regexpPattern:   cpuRegexpPattern,
+	}
+
+	err := forceOnlineResources(cpuResource, req.Cpus, onlineCPUMaxTries)
```

Comment:
Having functions `onlineCPUResources()` and `onlineMemResources()` seems more appropriate to make the code more readable.
```go
func onlineCPUResources(nbCPU int) error {
        resource := onlineResource{
		sysfsOnlinePath: sysfsCPUOnlinePath,
		regexpPattern:   cpuRegexpPattern,
	}

	var count uint32
	for i := uint32(0); i < onlineCPUMaxTries; i++ {
		r, err := onlineResources(resource)
		if err != nil {
			return err
		}
		count += r
		if count == nbCPU {
			return nil
		}
		time.Sleep(onlineCPUMemWaitTime)
	}

	return fmt.Errorf("only %d of %d were connected", count, expectedResources)
}

func onlineMemResources() error {
        resource := onlineResource{
		sysfsOnlinePath: sysfsMemOnlinePath,
		regexpPattern:   memRegexpPattern,
	}

        _, err := onlineResources(resource)
        return err
}
```